Fix virtlet crashing on domains listing in case of defined domains existence not registered in bolt

Bug was caught due to overlay affect of subsequent running integration tests and not fully cleaned env after failing tests:

```
I0518 14:30:15.316399   14216 manager.go:375] ListContainers: (*runtime.ListContainersRequest)(0xc420074550)({
 Filter: (*runtime.ContainerFilter)(<nil>)
})
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x5e9c56]

goroutine 37 [running]:
panic(0xa53b80, 0xc4200120c0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/Mirantis/virtlet/pkg/libvirttools.(*VirtualizationTool).getContainer(0xc42015a8c0, 0xc420074568, 0x3, 0x4, 0x0)
        /go/src/github.com/Mirantis/virtlet/pkg/libvirttools/virtualization.go:603 +0xb6
github.com/Mirantis/virtlet/pkg/libvirttools.(*VirtualizationTool).ListContainers(0xc42015a8c0, 0x0, 0x12, 0xc42004b9c0, 0x1, 0x1, 0xc4201c0101)
        /go/src/github.com/Mirantis/virtlet/pkg/libvirttools/virtualization.go:698 +0x16c
github.com/Mirantis/virtlet/pkg/manager.(*VirtletManager).ListContainers(0xc42015a980, 0x7f8966676160, 0xc4200dbc20, 0xc420074550, 0x0, 0xc42003fac8, 0x553858)
        /go/src/github.com/Mirantis/virtlet/pkg/manager/manager.go:376 +0x2ab
github.com/Mirantis/virtlet/vendor/k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime._RuntimeService_ListContainers_Handler(0xadf8a0, 0xc42015a980, 0x7f8966676160, 0xc4200dbc20, 0xc420085900, 0x0, 0x0, 0x0, 0x30, 0x0)
        /go/src/github.com/Mirantis/virtlet/vendor/k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime/api.pb.go:2462 +0x27d
github.com/Mirantis/virtlet/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc420186a00, 0x104c360, 0xc420016480, 0xc420127ef0, 0xc4201a3170, 0x103b790, 0xc4200dbbf0, 0x0, 0x0)
        /go/src/github.com/Mirantis/virtlet/vendor/google.golang.org/grpc/server.go:608 +0xc50
github.com/Mirantis/virtlet/vendor/google.golang.org/grpc.(*Server).handleStream(0xc420186a00, 0x104c360, 0xc420016480, 0xc420127ef0, 0xc4200dbbf0)
        /go/src/github.com/Mirantis/virtlet/vendor/google.golang.org/grpc/server.go:766 +0x6b0
github.com/Mirantis/virtlet/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4201c2680, 0xc420186a00, 0x104c360, 0xc420016480, 0xc420127ef0)
        /go/src/github.com/Mirantis/virtlet/vendor/google.golang.org/grpc/server.go:419 +0xab
created by github.com/Mirantis/virtlet/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /go/src/github.com/Mirantis/virtlet/vendor/google.golang.org/grpc/server.go:420 +0xa3
2017/05/18 14:30:15 transport: http2Client.notifyError got notified that the client transport was broken EOF.
2017/05/18 14:30:15 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/virtlet.sock: connect: connection refused"; Reconnecting to {"/tmp/virtlet.sock" <nil>}
--- FAIL: TestContainerStatuses (53.16s)
        container_test.go:114: Container created: "03ce3d01-dc9c-4e10-5afa-42e710ceccc9"
        container_test.go:132: Listing containers failure: rpc error: code = 13 desc = transport is closing
```

After failing volume test on ContainerCreate stage (https://travis-ci.org/Mirantis/virtlet/builds/233420469#L6529) defined domain remains in libvirt which causes virtlet crashing in subsequent test runs on list func with nil or only labels set filter each time (https://travis-ci.org/Mirantis/virtlet/builds/233420469#L8087)

ToDo:
- fix test cleanups
- fix virtlet listing func to just ignore and put warning log when it faces to "unknown" domains.

--- a/pkg/bolttools/virtualization.go
+++ b/pkg/bolttools/virtualization.go
@@ -172,7 +172,10 @@ func (b *BoltClient) GetContainerInfo(containerId string) (*metadata.ContainerIn
 
 		bucket := parentBucket.Bucket([]byte(containerId))
 		if bucket == nil {
-			// Container info removed, but sandbox still exists
+			// Can happen if:
+			// 1. Container info removed, but sandbox still exists
+			// 2. There's externally defined domain in libvirt daemon
+			//    using `virsh`or libvirt api, not registered in bolt
 			return nil
 		}
 

--- a/pkg/libvirttools/virtualization.go
+++ b/pkg/libvirttools/virtualization.go
@@ -603,6 +603,10 @@ func (v *VirtualizationTool) getContainer(domain *libvirt.Domain) (*kubeapi.Cont
 		return nil, err
 	}
 
+	if containerInfo == nil {
+		return nil, nil
+	}
+
 	podSandboxId := containerInfo.SandboxId
 
 	metadata := &kubeapi.ContainerMetadata{
@@ -701,6 +705,15 @@ func (v *VirtualizationTool) ListContainers(filter *kubeapi.ContainerFilter) ([]
 		container, err := v.getContainer(&domain)
 		if err != nil {
 			return nil, err
+		}
+
+		if container == nil {
+			containerId, err := v.tool.GetUUIDString(&domain)
+			if err != nil {
+				return nil, err
+			}
+			glog.V(0).Infof("Failed to find info in bolt for domain with id: %s, so just ignoring as not handled by virtlet.", containerId)
+			continue
 		}
 
 		if filterContainer(container, filter) {

--- a/tests/integration/container_create_test.go
+++ b/tests/integration/container_create_test.go
@@ -126,6 +126,11 @@ func TestContainerCreateStartListRemove(t *testing.T) {
 		ct.startContainer(createResp.ContainerId)
 	}
 
+	// Define external domain, i.e. not registered in bolt, to control virtlet performs well in that case
+	if err := defineDummyDomain(); err != nil {
+		t.Errorf("failed to define dummy domain to test List function: %v", err)
+	}
+
 	for _, tc := range []*containerFilterTestCase{
 		{
 			name:               "by container id",

--- a/tests/integration/utils.go
+++ b/tests/integration/utils.go
@@ -18,12 +18,18 @@ package integration
 
 import (
 	"fmt"
+	"log"
 	"os"
 	"time"
+
+	virtletutils "github.com/Mirantis/virtlet/pkg/utils"
+	libvirt "github.com/libvirt/libvirt-go"
+	libvirtxml "github.com/libvirt/libvirt-go-xml"
 )
 
 const (
-	maxTime = 60
+	maxTime    = 60
+	libvirtUri = "qemu+tcp://localhost/system"
 )
 
 func waitForSocket(filepath string) error {
@@ -35,4 +41,33 @@ func waitForSocket(filepath string) error {
 	}
 
 	return fmt.Errorf("Socket %s doesn't exist", filepath)
+}
+
+func defineDummyDomain() error {
+	conn, err := libvirt.NewConnect(libvirtUri)
+	if err != nil {
+		return err
+	}
+
+	uuid, err := virtletutils.NewUuid()
+	if err != nil {
+		log.Panicf("NewUuid(): %v", err)
+	}
+
+	domain := &libvirtxml.Domain{
+		Name: "dummy-" + uuid,
+		Type: "qemu",
+		OS: &libvirtxml.DomainOS{
+			Type: &libvirtxml.DomainOSType{Type: "hvm"},
+		},
+		Memory: &libvirtxml.DomainMemory{Value: 8192, Unit: "KiB"},
+	}
+
+	domainXML, err := domain.Marshal()
+	if err != nil {
+		log.Panicf("XML marshaling: %v", err)
+	}
+
+	_, err = conn.DomainDefineXML(domainXML)
+	return err
 }

